src/clients/python/experimental_api_v2/library/grpcclient.py

```diff
@@ -1174,3 +1231,158 @@ def get_response(self, as_json=False):
             return json.loads(MessageToJson(self._result))
         else:
             return self._result
+
+
+class InferStream:
+    """Supports sending inference requests and receiving corresponding
+    requests on a gRPC bi-directional stream.
+
+    Parameters
+    ----------
+    callback : function
+        Python function that is invoked upon receiving response from
+        the underlying stream. The function must reserve the last two
+        arguments (result, error) to hold InferResult and
+        InferenceServerException objects respectively which will be
+        provided to the function when executing the callback. The
+        ownership of these objects will be given to the user. The
+        'error' would be None for a successful inference.
+    """
+
+    def __init__(self, callback):
+        self._callback = callback
+        self._request_queue = queue.Queue()
+        self._handler = None
+        self._headers = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        """Gracefully close underlying gRPC streams. Note that this call
+        blocks till response of all currently enqueued requests are not
+        received.
+        """
+        if self._is_initialized():
+            self._request_queue.put(None)
+            if self._handler.is_alive():
+                self._handler.join()
+            self._handler = None
+
+    def set_headers(self, headers):
```

Comment:
OK, I'll merge this as is, but it seems to me that we should just have headers in the stream constructor just live we do for the infer methods. What is the use case for having a separate function?